api/v1alpha1/condition.go

```diff
@@ -0,0 +1,181 @@
+// +groupName=core.kubebb.k8s.com.cn
+package v1alpha1
+
+import (
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// A ConditionType represents a condition a resource could be in.
+type ConditionType string
+
+// Some common Condition types.
+const (
+	// TypeReady resources are believed to be ready to handle work.
+	TypeReady ConditionType = "Ready"
+
+	// TypeSynced resources are believed to be in sync with the
+	// Kubernetes resources that manage their lifecycle.
+	TypeSynced ConditionType = "Synced"
+)
+
+// A ConditionReason represents the reason a resource is in a condition.
+// Should be only one word
+type ConditionReason string
+
+// Some common Condition resons.
+const (
+	ReasonUnavailable      ConditionReason = "Unavailable"
+	ReasonCreating         ConditionReason = "Creating"
+	ReasonDeleting         ConditionReason = "Deleting"
+	ReasonReconcileSuccess ConditionReason = "ReconcileSuccess"
+	ReasonReconcileError   ConditionReason = "ReconcileError"
+	ReasonReconcilePaused  ConditionReason = "ReconcilePaused"
+)
+
+// A Condition that may apply to a resource.
+type Condition struct {
+	// Type of this condition. At most one of each condition type may apply to
+	// a resource at any point in time.
+	Type ConditionType `json:"type"`
+
+	// Status of this condition; is it currently True, False, or Unknown
+	Status corev1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the last time this condition transitioned from one
+	// status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// A Reason for this condition's last transition from one status to another.
+	Reason ConditionReason `json:"reason"`
+
+	// A Message containing details about this condition's last transition from
+	// one status to another, if any.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// Equal returns true if the condition is identical to the supplied condition,
+// ignoring the LastTransitionTime.
+func (c Condition) Equal(other Condition) bool {
+	return c.Type == other.Type &&
+		c.Status == other.Status &&
+		c.Reason == other.Reason &&
+		c.Message == other.Message
+}
+
+// WithMessage returns a condition by adding the provided message to existing
+// condition.
+func (c Condition) WithMessage(msg string) Condition {
+	c.Message = msg
+	return c
+}
+
+// NOTE: Conditions are implemented as a slice rather than a map to comply
+// with Kubernetes API conventions. Ideally we'd comply by using a map that
+// marshalled to a JSON array, but doing so confuses the CRD schema generator.
+// https://github.com/kubernetes/community/blob/9bf8cd/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
+
+// NOTE: Do not manipulate Conditions directly. Use the Set method.
+
+// A ConditionedStatus reflects the observed status of a resource. Only
+// one condition of each type may exist.
+type ConditionedStatus struct {
```

Comment:
I think we can reuse this struct and just set different `ConditionType` and `ConditionReason` depending on the CRD

Reply:
Looks good for now. We can use this as common condition typed struct.  @0xff-dev  any thoughts?

Reply:
Agree.